test/src/main/java/org/apache/accumulo/test/fate/zookeeper/ZooLockIT.java

```diff
@@ -306,6 +363,245 @@ public void testUnexpectedEvent() throws Exception {
+    public void run() {
+      ConnectedWatcher watcher = new ConnectedWatcher();
+      try {
+        try (ZooKeeperWrapper zk =
+            new ZooKeeperWrapper(getCluster().getZooKeepers(), 30000, watcher)) {
+          zk.addAuthInfo("digest", "accumulo:secret".getBytes(UTF_8));
+          while (!watcher.isConnected()) {
+            Thread.sleep(50);
+          }
+          ZooReaderWriter zrw = new ZooReaderWriter(getCluster().getZooKeepers(), 30000, "secret") {
+            @Override
+            public ZooKeeper getZooKeeper() {
+              return zk;
+            }
+          };
+          ZooLock zl = new ZooLock(zrw, parent);
+          zl.setVMLockPrefix(new Prefix(name));
+          getLockLatch.countDown(); // signal we are done
+          getLockLatch.await(); // wait for others to finish
+          zl.lock(lockWatcher, "test1".getBytes(UTF_8)); // race to the lock
+          lockCompletedLatch.countDown();
+          unlockLatch.await();
+          zl.unlock();
+        }
+      } catch (IOException | InterruptedException | KeeperException e) {
+        throw new RuntimeException(e);
+      }
+    }
+
+    @Override
+    public String toString() {
+      return "LockWorker [name=" + name + ", holdsLock()=" + holdsLock() + "]";
+    }
+
+  }
+
+  private int parseLockWorkerName(String child) {
+    if (child.startsWith("zlock#00000000-0000-0000-0000-000000000000#")) {
+      return 0;
+    } else if (child.startsWith("zlock#00000000-0000-0000-0000-111111111111#")) {
+      return 1;
+    } else if (child.startsWith("zlock#00000000-0000-0000-0000-222222222222#")) {
+      return 2;
+    } else if (child.startsWith("zlock#00000000-0000-0000-0000-333333333333#")) {
+      return 3;
+    } else {
+      return -1;
+    }
+  }
+
+  @Test(timeout = 60000)
+  public void testLockParallel() throws Exception {
+    String parent = "/zlParallel";
+
+    ConnectedWatcher watcher = new ConnectedWatcher();
+    try (ZooKeeperWrapper zk = new ZooKeeperWrapper(getCluster().getZooKeepers(), 30000, watcher)) {
+      zk.addAuthInfo("digest", "accumulo:secret".getBytes(UTF_8));
+
+      while (!watcher.isConnected()) {
+        Thread.sleep(50);
+      }
+      // Create the parent node
+      zk.createOnce(parent, new byte[0], ZooDefs.Ids.OPEN_ACL_UNSAFE, CreateMode.PERSISTENT);
+
+      int numWorkers = 4;
+      final CountDownLatch getLockLatch = new CountDownLatch(numWorkers);
+      final CountDownLatch lockFinishedLatch = new CountDownLatch(numWorkers);
+      final List<LockWorker> workers = new ArrayList<>(numWorkers);
+      final List<Thread> threads = new ArrayList<>(numWorkers);
+      for (int i = 0; i < numWorkers; i++) {
+        String name =
+            "zlock#00000000-0000-0000-0000-AAAAAAAAAAAA#".replaceAll("A", Integer.toString(i));
+        LockWorker w = new LockWorker(parent, name, getLockLatch, lockFinishedLatch);
+        Thread t = new Thread(w);
+        workers.add(w);
+        threads.add(t);
+        t.start();
```

Comment:
Not sure what will happen w/ this test if an exception happens in the background thread.  Could refactor the test to use an executor and futures.  Futures are nice because they make it easy to know when a background task had an exception.

Reply:
I implemented a check, but not using Futures.